### PR TITLE
Fix builds to work with package references

### PR DESCRIPTION
**Pull request by weshaggard**
After we switched to package references some builds did not
work correctly because there was a mix of package and project
reference in the resolution graph. Note this only repro's if
/p:SkipDevBuildNumber=true is set which caused the package and
the project to have the exact same version number but one coming
from source and the other from nuget.

To fix this I had to change Azure.Core.Test project to take a
project reference to keyvault.

Also needed to update the eventhubs checkpoint test project to
remove the project dependency on the eventhubs test project which
caused a mix-match of package/project references for eventhubs.

cc @pakrym @jsquire @chidozieononiwu @danieljurek

### CODE REVIEW

File: sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj
Comment (pakrym) [inline sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj]: > To fix this I had to change Azure.Core.Test project to take a
project reference to keyvault.

Hm?

Why does this help? It stil results in Azure.Core package reference of version 1.0.0.

Comment (weshaggard) [reply sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj]: Good point I'll keep digging to see if I can figure that out. For reference here are the errors we get without this change:
```
                   "D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj" (Restore target) (1) ->
                   (Restore target) -> 
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Threading.Tasks.Extensions': System.Threading.Tasks.Extensions (>= 4.5.2) (via package/Azure.Identity 1.0.0), System.Threading.Tasks.Extensions (>= 4.5.2) (via package/Microsoft.Bcl.AsyncInterfaces 1.0.0), System.Threading.Tasks.Extensions (>= 4.5.1) (via package/Moq 4.10.1), System.Threading.Tasks.Extensions (>= 4.5.2) (via package/System.Text.Json 4.6.0), System.Threading.Tasks.Extensions (>= 4.5.2) (via project/Azure.Core 1.0.0), System.Threading.Tasks.Extensions (>= 4.5.2) (via project/Azure.Security.KeyVault.Secrets 4.0.0), System.Threading.Tasks.Extensions (>= 4.5.1) (via package/Microsoft.AspNetCore.Server.Kestrel.Core 2.2.0), System.Threading.Tasks.Extensions (>= 4.5.1) (via package/System.IO.Pipelines 4.5.2) Framework: (.NETFramework,Version=v4.6.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Buffers': System.Buffers (>= 4.4.0) (via package/System.Memory 4.5.3), System.Buffers (>= 4.5.0) (via package/System.Text.Json 4.6.0), System.Buffers (>= 4.5.0) (via project/Azure.Core 1.0.0), System.Buffers (>= 4.5.0) (via package/Azure.Core 1.0.0), System.Buffers (>= 4.5.0) (via package/Microsoft.Net.Http.Headers 2.2.0), System.Buffers (>= 4.5.0) (via package/Microsoft.AspNetCore.Http.Extensions 2.2.0), System.Buffers (>= 4.5.0) (via package/Microsoft.Net.Http.Headers 2.2.0), System.Buffers (>= 4.4.0) (via package/System.IO.Pipelines 4.5.2), System.Buffers (>= 4.4.0) (via package/System.IO.Pipelines 4.5.2) Framework: (.NETFramework,Version=v4.6.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Numerics.Vectors': System.Numerics.Vectors (>= 4.4.0) (via package/System.Memory 4.5.3), System.Numerics.Vectors (>= 4.5.0) (via package/System.Text.Json 4.6.0), System.Numerics.Vectors (>= 4.5.0) (via project/Azure.Core 1.0.0), System.Numerics.Vectors (>= 4.5.0) (via package/Microsoft.AspNetCore.Server.Kestrel.Core 2.2.0), System.Numerics.Vectors (>= 4.5.0) (via package/Azure.Core 1.0.0) Framework: (.NETFramework,Version=v4.6.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Runtime.CompilerServices.Unsafe': System.Runtime.CompilerServices.Unsafe (>= 4.5.2) (via package/System.Memory 4.5.3), System.Runtime.CompilerServices.Unsafe (>= 4.6.0) (via package/System.Text.Json 4.6.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.2) (via package/System.Threading.Tasks.Extensions 4.5.2), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.AspNetCore.Server.Kestrel.Core 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.2) (via package/System.Threading.Tasks.Extensions 4.5.2), System.Runtime.CompilerServices.Unsafe (>= 4.5.0) (via package/System.Threading.Tasks.Extensions 4.5.1), System.Runtime.CompilerServices.Unsafe (>= 4.5.2) (via package/System.Threading.Tasks.Extensions 4.5.2), System.Runtime.CompilerServices.Unsafe (>= 4.5.2) (via package/System.Threading.Tasks.Extensions 4.5.2), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.0) (via package/System.Threading.Tasks.Extensions 4.5.1), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0), System.Runtime.CompilerServices.Unsafe (>= 4.5.1) (via package/Microsoft.Extensions.Primitives 2.2.0) Framework: (.NETFramework,Version=v4.6.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Diagnostics.DiagnosticSource': System.Diagnostics.DiagnosticSource (>= 4.6.0) (via project/Azure.Core 1.0.0), System.Diagnostics.DiagnosticSource (>= 4.5.0) (via package/Microsoft.AspNetCore.Hosting 2.2.0), System.Diagnostics.DiagnosticSource (>= 4.6.0) (via package/Azure.Core 1.0.0) Framework: (.NETFramework,Version=v4.6.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Threading.Tasks.Extensions': System.Threading.Tasks.Extensions (>= 4.5.2) (via package/Azure.Identity 1.0.0), System.Threading.Tasks.Extensions (>= 4.5.1) (via package/Moq 4.10.1), System.Threading.Tasks.Extensions (>= 4.5.2) (via project/Azure.Core 1.0.0), System.Threading.Tasks.Extensions (>= 4.5.2) (via project/Azure.Security.KeyVault.Secrets 4.0.0), System.Threading.Tasks.Extensions (>= 4.5.1) (via package/Microsoft.AspNetCore.Server.Kestrel.Core 2.2.0), System.Threading.Tasks.Extensions (>= 4.3.0) (via package/System.Xml.ReaderWriter 4.3.0), System.Threading.Tasks.Extensions (>= 4.3.0) (via package/System.Xml.ReaderWriter 4.3.0) Framework: (.NETCoreApp,Version=v2.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Buffers': System.Buffers (>= 4.5.0) (via project/Azure.Core 1.0.0), System.Buffers (>= 4.5.0) (via package/Azure.Core 1.0.0), System.Buffers (>= 4.5.0) (via package/Microsoft.Net.Http.Headers 2.2.0), System.Buffers (>= 4.5.0) (via package/Microsoft.AspNetCore.Http.Extensions 2.2.0), System.Buffers (>= 4.5.0) (via package/Microsoft.Net.Http.Headers 2.2.0) Framework: (.NETCoreApp,Version=v2.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Diagnostics.DiagnosticSource': System.Diagnostics.DiagnosticSource (>= 4.6.0) (via project/Azure.Core 1.0.0), System.Diagnostics.DiagnosticSource (>= 4.5.0) (via package/Microsoft.AspNetCore.Hosting 2.2.0), System.Diagnostics.DiagnosticSource (>= 4.6.0) (via package/Azure.Core 1.0.0) Framework: (.NETCoreApp,Version=v2.1)
                     D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj : error NU1106: Unable to satisfy conflicting requests for 'System.Numerics.Vectors': System.Numerics.Vectors (>= 4.5.0) (via project/Azure.Core 1.0.0), System.Numerics.Vectors (>= 4.5.0) (via package/Microsoft.AspNetCore.Server.Kestrel.Core 2.2.0), System.Numerics.Vectors (>= 4.5.0) (via package/Azure.Core 1.0.0) Framework: (.NETCoreApp,Version=v2.1)
                     c:\program files\dotnet\sdk\3.0.100\NuGet.targets(123,5): error : An item with the same key has already been added. Key: (Azure.Core, 1.0.0) [D:\adp\azure-sdk-for-net\sdk\core\Azure.Core\tests\Azure.Core.Tests.csproj]
```

Comment (pakrym) [reply sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj]: We may need to file `An item with the same key has already been added.` on nuget, it's a `dict.Add` exception not something customers should see.

File: sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: Would you please be so kind as to rename the folder to `SharedSource` to match the convention used with other cross-project sharing within the Event Hubs ecosystem?

**Review by jsquire — APPROVED** [review]
I'd really like to see the `SharedSource` areas of Event Hubs move to a designated area outside of the scope of a specific project, so that it is immediately clear to folks working on things that it is shared between different projects.

Our current repository structure is not friendly to that; we're optimized for things that do not cross package boundaries, which results in having a project "own" the shared areas and consumers reaching into that space to borrow things.   In my opinion, this makes it more difficult to reason about what can safely be changed locally versus what changes cascade to other projects and to which ones,  thus weakening the package isolation.   

Any ideas on how we could potentially include a "We share this across all of Event Hubs land"  or "We share this across all things on the SDK repostiory" area to help clarify usage scope without running afoul of the intended repository structure?

**Comment by weshaggard** [comment]
>Any ideas on how we could potentially include a "We share this across all of Event Hubs land" or "We share this across all things on the SDK repostiory" area to help clarify usage scope without running afoul of the intended repository structure?

I agree we have been missing some common place to share source code across projects. That is part of the reason for my changes to the repo structure in https://github.com/Azure/azure-sdk/pull/733, I expect us to create a root level folder called common that we can start to put these shared source files under. I will update this PR to match that convention.

File: sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj
Comment (weshaggard) [reply sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: What do you think the "Source" adds to the name? It seems kind of redundant to me. I will think about the name a little more and try to come up with one that makes since given the fact I'm going to move it to the common folder.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: Do you think that we could do that in a different context, and preferably after GA for Event Hubs?   I'm blocked on release at present and looking at a schedule for GA delivery with very, very little room.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: I like "source" in this case to denote that this is shared as a source file and to imply the semantics of "you change this here, and other things change too."    I'm happy to discuss naming at some point, but I'd like to push this until after we GA in December due to tight timelines.

For now, since we'll be local to Event Hubs, I'd like to maintain control of the naming as the library owner.  I'm happy to discuss changes to it at the point we move it outside of the Event Hubs scope.

**Comment by pakrym** [comment]
> I agree we have been missing some common place to share source code across projects. That is part of the reason for my changes to the repo structure in Azure/azure-sdk#733, I expect us to create a root level folder called common that we can start to put these shared source files under. I will update this PR to match that convention.

Are we going to put EvenHubs test shared code into root `common` direcotry? Seems overkill.

**Comment by jsquire** [comment]
> > I agree we have been missing some common place to share source code across projects. That is part of the reason for my changes to the repo structure in [Azure/azure-sdk#733](https://github.com/Azure/azure-sdk/pull/733), I expect us to create a root level folder called common that we can start to put these shared source files under. I will update this PR to match that convention.
> 
> Are we going to put EvenHubs test shared code into root `common` direcotry? Seems overkill.

I'd agree, since we're sharing only within the Event Hubs context and do not intend to make it something that we want other client libraries to grab.   I was thinking of something more local to the `/sdk/eventhub` area.    That said, I really can't afford to bog down at this point, so I'd like to ask that we please table this for now and I'll make whatever local tweaks within the Event Hubs project.

File: sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: Just a note that I'm fine with "Shared" being the folder in the tests library that is doing the sharing.  It's the destination that is borrowing things that I'd like to ensure that semantics are clear for.

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: ... what I'm asking for is this, if you would:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <AssemblyName>Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests</AssemblyName>
    <VersionPrefix>1.0.0</VersionPrefix>
    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
    <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
    <ExcludeRecordingFramework>true</ExcludeRecordingFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" />
    <PackageReference Include="System.Net.WebSockets.Client" />
    <PackageReference Include="System.ValueTuple" />
    <PackageReference Include="NUnit" />
    <PackageReference Include="NUnit3TestAdapter" />
    <PackageReference Include="Moq" />
    <PackageReference Include="Polly" />
    <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" />
    <PackageReference Include="Microsoft.Azure.Management.EventHub" />
    <PackageReference Include="System.Threading.Tasks.Extensions" />
    <PackageReference Include="Microsoft.Azure.Management.Storage" />
    <PackageReference Include="Microsoft.Azure.Management.ResourceManager" />
    <PackageReference Include="Microsoft.Azure.Services.AppAuthentication" />
  </ItemGroup>

  <ItemGroup>
    <ProjectReference Include="..\src\Azure.Messaging.EventHubs.CheckpointStore.Blobs.csproj" />
    <ProjectReference Include="..\samples\Azure.Messaging.EventHubs.CheckpointStore.Blobs.Samples.csproj" />
  </ItemGroup>

  <!-- Import the shared infrastructure from the Event Hubs client tests (shared source) -->
  <ItemGroup>
    <Compile Include="..\..\Azure.Messaging.EventHubs\tests\Shared\*.cs" Link="SharedSource\%(Filename)%(Extension)" />
  </ItemGroup>
</Project>
```

File: sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj
Comment (weshaggard) [reply sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj]: @pakrm looks like we are hitting https://github.com/NuGet/Home/issues/6795. This fixes it though and so I'll keep this fix.

File: sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj
Comment (weshaggard) [reply sdk/eventhub/Azure.Messaging.EventHubs.CheckpointStore.Blobs/tests/Azure.Messaging.EventHubs.CheckpointStore.Blobs.Tests.csproj]: Thanks for the clarification I didn't realize you were meaning the folder view in VS. I updated to use your sample.

**Comment by weshaggard** [comment]
I've updated this PR and once it is green we can merge.

**Comment by weshaggard** [comment]
>Are we going to put EvenHubs test shared code into root common direcotry? Seems overkill.

If it is shared between 2 projects then I'm fine with the way we are doing it now, and I would like to stick to the structure so we can maintain the folder structure under sdk folder (i.e. `<service>\<package>`). For things were we want to share it more broadly we should put it under a root common folder and name it to match what it provides. Then folks can import it or not depending on there needs.

**Comment by weshaggard** [comment]
EventHubs leg failed with a timeout on linux. I filed https://github.com/Azure/azure-sdk-for-net/issues/8618 for that as it seems to have happened in a few other runs as well.

**Pull request merged**
